File: tests/vectors/integration/embedded_choice/test_embedded_choice.py
"""Integration tests for embedded CHOICE with AUTOMATIC TAGS.

Tests embedded CHOICE with AUTOMATIC TAGS, extension markers.

Adapted from vlm/asn1c tests-c-compiler/67-embedded-choice-OK.asn1 (MIT license).
https://github.com/vlm/asn1c
"""
from __future__ import annotations

import os
import sys

_integ_output = os.environ.get("INTEG_OUTPUT_DIR")
if _integ_output:
    sys.path.insert(0, _integ_output)

import pytest

try:
    from EmbeddedChoiceModule import Container, EmbeddedChoice, ExtendedSequence
    HAS_GENERATED = True
except ImportError:
    HAS_GENERATED = False


@pytest.mark.skipif(not HAS_GENERATED, reason="Generated types not available")
class TestEmbeddedChoice:
    """Test embedded CHOICE with AUTOMATIC TAGS encode/decode roundtrips."""

    def test_choice_integer(self):
        """Test EmbeddedChoice selecting integer alternative."""
        t = EmbeddedChoice(integer=42)
        encoded = t.encode_der()
        decoded = EmbeddedChoice.decode_der(encoded)
        assert decoded.integer == 42

    def test_choice_string(self):
        """Test EmbeddedChoice selecting string alternative."""
        t = EmbeddedChoice(string="hello")
        encoded = t.encode_der()
        decoded = EmbeddedChoice.decode_der(encoded)
        assert decoded.string == "hello"

    def test_choice_octet(self):
        """Test EmbeddedChoice selecting octet string alternative."""
        t = EmbeddedChoice(octet=b"\x01\x02\x03")
        encoded = t.encode_der()
        decoded = EmbeddedChoice.decode_der(encoded)
        assert decoded.octet == b"\x01\x02\x03"

    def test_container_with_choice_integer(self):
        """Test Container with EmbeddedChoice (integer)."""
        choice = EmbeddedChoice(integer=100)
        c = Container(id=1, value=choice)
        encoded = c.encode_der()
        decoded = Container.decode_der(encoded)
        assert decoded.id == 1
        assert decoded.value.integer == 100

    def test_container_with_choice_string(self):
        """Test Container with EmbeddedChoice (string)."""
        choice = EmbeddedChoice(string="test data")
        c = Container(id=2, value=choice)
        encoded = c.encode_der()
        decoded = Container.decode_der(encoded)
        assert decoded.id == 2
        assert decoded.value.string == "test data"

    def test_container_with_choice_octet(self):
        """Test Container with EmbeddedChoice (octet string)."""
        choice = EmbeddedChoice(octet=b"\xDE\xAD\xBE\xEF")
        c = Container(id=3, value=choice)
        encoded = c.encode_der()
        decoded = Container.decode_der(encoded)
        assert decoded.id == 3
        assert decoded.value.octet == b"\xDE\xAD\xBE\xEF"

    def test_extended_sequence_base(self):
        """Test ExtendedSequence with only base fields."""
        t = ExtendedSequence(name="base", value=1)
        encoded = t.encode_der()
        decoded = ExtendedSequence.decode_der(encoded)
        assert decoded.name == "base"
        assert decoded.value == 1

    def test_extended_sequence_with_optional(self):
        """Test ExtendedSequence with optional extension field."""
        t = ExtendedSequence(name="extended", value=2, extra="bonus")
        encoded = t.encode_der()
        decoded = ExtendedSequence.decode_der(encoded)
        assert decoded.name == "extended"
        assert decoded.value == 2
        assert decoded.extra == "bonus"

    def test_choice_tag_uniqueness(self):
        """Verify that different CHOICE alternatives produce different tag bytes."""
        t_int = EmbeddedChoice(integer=1)
        t_str = EmbeddedChoice(string="x")
        encoded_int = t_int.encode_der()
        encoded_str = t_str.encode_der()
        # Different alternatives should have different first byte (tag)
        assert encoded_int[0] != encoded_str[0], (
            "CHOICE alternatives should have distinct tags with AUTOMATIC TAGS"
        )

    def test_container_list_of_choices(self):
        """Test Container with list of choices."""
        choices = [
            EmbeddedChoice(integer=1),
            EmbeddedChoice(string="two"),
            EmbeddedChoice(octet=b"\x03"),
        ]
        c = Container(id=10, value=choices[0])
        encoded = c.encode_der()
        decoded = Container.decode_der(encoded)
        assert decoded.id == 10
        assert decoded.value.integer == 1

File: tests/vectors/embedded_choice/test_embedded_choice.rs
//! ```cargo
//! [dependencies]
//! asnvil-runtime-rust = { path = "__REPO_ROOT__/asnvil-runtime-rust" }
//! num-bigint = "0.4"
//! ```

#[path = "./EmbeddedChoiceModule.rs"]
mod generated;

use generated::{Container, EmbeddedChoice, ExtendedSequence};
use num_bigint::BigInt;

#[test]
fn test_choice_integer() {
    let t = EmbeddedChoice::integer(BigInt::from(42));
    let encoded = t.encode_der().expect("Failed to encode");
    let decoded = EmbeddedChoice::decode_der(&encoded).expect("Failed to decode");
    match decoded {
        EmbeddedChoice::integer(v) => assert_eq!(v, BigInt::from(42)),
        _ => panic!("Expected integer"),
    }
}

#[test]
fn test_choice_string() {
    let t = EmbeddedChoice::string("hello".to_string());
    let encoded = t.encode_der().expect("Failed to encode");
    let decoded = EmbeddedChoice::decode_der(&encoded).expect("Failed to decode");
    match decoded {
        EmbeddedChoice::string(v) => assert_eq!(v, "hello"),
        _ => panic!("Expected string"),
    }
}

#[test]
fn test_choice_octet() {
    let t = EmbeddedChoice::octet(vec![1, 2, 3]);
    let encoded = t.encode_der().expect("Failed to encode");
    let decoded = EmbeddedChoice::decode_der(&encoded).expect("Failed to decode");
    match decoded {
        EmbeddedChoice::octet(v) => assert_eq!(v, vec![1, 2, 3]),
        _ => panic!("Expected octet"),
    }
}

#[test]
fn test_container_with_choice_integer() {
    let choice = EmbeddedChoice::integer(BigInt::from(100));
    let c = Container { id: BigInt::from(1), value: choice };
    let encoded = c.encode_der().expect("Failed to encode");
    let decoded = Container::decode_der(&encoded).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(1));
    match decoded.value {
        EmbeddedChoice::integer(v) => assert_eq!(v, BigInt::from(100)),
        _ => panic!("Expected integer"),
    }
}

#[test]
fn test_container_with_choice_string() {
    let choice = EmbeddedChoice::string("test data".to_string());
    let c = Container { id: BigInt::from(2), value: choice };
    let encoded = c.encode_der().expect("Failed to encode");
    let decoded = Container::decode_der(&encoded).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(2));
    match decoded.value {
        EmbeddedChoice::string(v) => assert_eq!(v, "test data"),
        _ => panic!("Expected string"),
    }
}

#[test]
fn test_container_with_choice_octet() {
    let choice = EmbeddedChoice::octet(vec![0xDE, 0xAD, 0xBE, 0xEF]);
    let c = Container { id: BigInt::from(3), value: choice };
    let encoded = c.encode_der().expect("Failed to encode");
    let decoded = Container::decode_der(&encoded).expect("Failed to decode");
    assert_eq!(decoded.id, BigInt::from(3));
    match decoded.value {
        EmbeddedChoice::octet(v) => assert_eq!(v, vec![0xDE, 0xAD, 0xBE, 0xEF]),
        _ => panic!("Expected octet"),
    }
}

#[test]
fn test_extended_sequence_base() {
    let t = ExtendedSequence { name: "base".to_string(), value: BigInt::from(1), extra: None };
    let encoded = t.encode_der().expect("Failed to encode");
    let decoded = ExtendedSequence::decode_der(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "base");
    assert_eq!(decoded.value, BigInt::from(1));
}

#[test]
fn test_extended_sequence_with_optional() {
    let t = ExtendedSequence { name: "extended".to_string(), value: BigInt::from(2), extra: Some("bonus".to_string()) };
    let encoded = t.encode_der().expect("Failed to encode");
    let decoded = ExtendedSequence::decode_der(&encoded).expect("Failed to decode");
    assert_eq!(decoded.name, "extended");
    assert_eq!(decoded.value, BigInt::from(2));
    assert_eq!(decoded.extra, Some("bonus".to_string()));
}

#[test]
fn test_choice_tag_uniqueness() {
    let t_int = EmbeddedChoice::integer(BigInt::from(1));
    let t_str = EmbeddedChoice::string("x".to_string());
    let encoded_int = t_int.encode_der().expect("Failed to encode");
    let encoded_str = t_str.encode_der().expect("Failed to encode");
    assert_ne!(encoded_int[0], encoded_str[0], "CHOICE alternatives should have distinct tags");
}